src/Control/Algebra.hs

```diff
@@ -131,46 +131,46 @@ send sig = runIdentity <$> alg (fmap Identity . runIdentity) (inj sig) (Identity
 -- base
 
 instance Algebra (Lift IO) IO where
-  alg hdl (LiftWith with k) ctx = with hdl ctx >>= hdl . fmap k
+  alg hdl (LiftWith with) = with hdl
 
 instance Algebra (Lift Identity) Identity where
-  alg hdl (LiftWith with k) ctx = with hdl ctx >>= hdl . fmap k
+  alg hdl (LiftWith with) = with hdl
 
 instance Algebra Choose NonEmpty where
-  alg hdl (Choose m) ctx = hdl (m True <$ ctx) S.<> hdl (m False <$ ctx)
+  alg _ Choose ctx = (True <$ ctx) :| [ False <$ ctx ]
 
 instance Algebra Empty Maybe where
   alg _ Empty _ = Nothing
 
 instance Algebra (Error e) (Either e) where
   alg hdl sig ctx = case sig of
-    L (Throw e)     -> Left e
-    R (Catch m h k) -> either (hdl . (<$ ctx) . h) pure (hdl (m <$ ctx)) >>= hdl . fmap k
+    L (Throw e)   -> Left e
+    R (Catch m h) -> either (hdl . (<$ ctx) . h) pure (hdl (m <$ ctx))
 
 instance Algebra (Reader r) ((->) r) where
   alg hdl sig ctx = case sig of
-    Ask       k -> \ r -> hdl (k r <$ ctx) r
-    Local f m k -> \ r -> hdl (fmap k (hdl (m <$ ctx) (f r))) r
+    Ask       -> (<$ ctx)
+    Local f m -> hdl (m <$ ctx) . f
 
 instance Algebra NonDet [] where
-  alg hdl sig ctx = case sig of
-    L Empty      -> []
-    R (Choose k) -> hdl (k True <$ ctx) ++ hdl (k False <$ ctx)
+  alg _ sig ctx = case sig of
+    L Empty  -> []
+    R Choose -> [ True <$ ctx, False <$ ctx ]
```

Comment:
This is one of the best examples of how cool this change turned out to be. Lacking a continuation means we have to `return` a value instead of calling `k` with it, and that value is `ctx a` (whatever `a` is for that constructor). Since `Choose :: Choose m Bool`, that means we need to return `[ctx Bool]`. Hence, instead of concatenating the lists produced by the continuations, we produce the two-element list above and let the `Monad` instance for `[]` take care of the rest externally to `alg`.

All in all, this change should result in far fewer binds being issued. Many of these would ideally have been compiled away anyhow, but there’s now that much less work for the simplifier to do. And since the algebras themselves are (often dramatically!) simpler as a result, it’s win-win-win.